src/resolver.js

```
@@ -1,135 +1,70 @@
+    if (CID.isCID(node)) {
+      return {
         value: node,
-        remainderPath: ''
-      })
-    }
-
-    // within scope
-
-    const parts = path.split('/')
-    const val = traverse(node).get(parts)
-
-    if (val !== undefined) {
-      return callback(null, {
-        value: val,
-        remainderPath: ''
-      })
-    }
-
-    // out of scope
-    let value
-    const len = parts.length
-
-    for (let i = 0; i < len; i++) {
-      const partialPath = parts.shift()
-
-      if (Array.isArray(node) && !Buffer.isBuffer(node)) {
-        value = node[Number(partialPath)]
-      } if (node[partialPath]) {
-        value = node[partialPath]
-      } else {
-        // can't traverse more
-        if (!value) {
-          return callback(new Error('path not available at root'))
-        } else {
-          parts.unshift(partialPath)
-          return callback(null, {
-            value: value,
-            remainderPath: parts.join('/')
-          })
-        }
+        remainderPath: parts.join('/')
       }
-      node = value
     }
-  })
-}
-
-function flattenObject (obj, delimiter) {
-  delimiter = delimiter || '/'
-
-  if (Object.keys(obj).length === 0) {
-    return []
   }
 
-  return traverse(obj).reduce(function (acc, x) {
-    if (CID.isCID(x)) {
-      this.update(undefined)
-    }
-    const path = this.path.join(delimiter)
-
-    if (path !== '') {
-      acc.push({ path: path, value: x })
-    }
-    return acc
-  }, [])
+  return {
+    value: node,
+    remainderPath: ''
+  }
 }
 
-/*
- * tree: returns a flattened array with paths: values of the project. options
- * are option (i.e. nestness)
- */
-exports.tree = (binaryBlob, options, callback) => {
-  if (typeof options === 'function') {
-    callback = options
-    options = undefined
+const traverse = function * (node, path) {
+  // Traverse only objects and arrays
+  if (Buffer.isBuffer(node) || CID.isCID(node) || typeof node === 'string' ||
+      node === null) {
```

the original suggestion was to change from `node === undefined` to `node == null` because the `==` lets you match both `undefined` and `null` in one shot, the strict `=== null` now leaves out the `undefined` case which I think might still be important?

---

This was intentional. The object should comply to the IPLD Data Model, where only `null` is a thing, but `undefined` is not. So it will fail early.

---

is it impossible to encode a block with an `undefined` property? because "fail" in this case is going to be an `Error` raised by `Object.keys()` below which might be surprising

---

I think so. I'd like to try it as least (as a strong believer of failing early).